File: solutions/c/binary-search/2/binary_search.c
#include "binary_search.h"

int *binary_search(int value, const int *arr, size_t length) {
    int left = 0;
    int right = length - 1;

    if (length == 0 || arr == NULL) {
        return NULL;
    }

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (value == arr[mid])
            return (int*)&arr[mid];

        if (value > arr[mid]) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return NULL;
}

File: solutions/typescript/binary-search/1/binary-search.ts
export function find(haystack: number[], needle: number): number | never {
    // Implement the binary search algorithm to find the index of the needle in the haystack
    //
    let left = 0;
    let right = haystack.length - 1;
    while (left <= right) {
        const mid = left + Math.floor((right - left) / 2);
        if (haystack[mid] === needle) {
            return mid;
        }
        if (haystack[mid] < needle) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    throw new Error('Value not in array');
}
